[CS2113T-W09-4] PatientManager

brandonfoong opened the pull request:
PatientManager helps General Practioners in polyclinics to better manage the medical records of their patients. It is optimized for CLI users so that frequent tasks can be done faster by typing in commands.

leowxx committed:
Modified Storage and StorageTest

leowxx committed:
Revert "Modified Storage and StorageTest"

This reverts commit ccbd5b2f9323572e5e76ccb703ecae893d779909.

Emkay16 committed:
Merge branch 'master' into Mabel-userGuide

fsgmhoward committed:
Add saveFile() method for Data class and multiple improvements

- Change patients in Data to private and fix StorageTest accordingly
- Add/modify constructor for Data class
- Add skeleton for StorageException
- Fix warnings and unused imports
- Replace generic Exception with InvalidInputException for ParserTest
- Some minor changes on exception messages

Emkay16 committed:
Add example outputs and FAQ section

leowxx committed:
Merge branch 'master' of https://github.com/AY2021S2-CS2113T-W09-4/tp

# Conflicts:
#	src/main/java/seedu/duke/Storage.java
#	src/test/java/seedu/duke/StorageTest.java

brandonfoong committed:
Add details and diagrams for the Logic and Model components

brandonfoong committed:
Add more details on the Logic section

brandonfoong committed:
Fix formatting errors

Emkay16 committed:
Add splitStringIntoArrayList method

brandonfoong committed:
Update _config.yml

brandonfoong committed:
Minor edits

brandonfoong committed:
Merge branch 'Brandon-Documentation' of github.com:brandonfoong/tp into Brandon-Documentation

Emkay16 committed:
Add loadData test

fsgmhoward committed:
Merge branch 'master' into kaihan-DeleteCommand

fsgmhoward committed:
Merge pull request #68 from hiongkaihan/kaihan-DeleteCommand

Add deleteCommand Function, JunitTest for deleteCommand and NRIC checkum

Emkay16 committed:
Update code to be compliant with coding standard

leowxx committed:
Merge branch 'master' of https://github.com/AY2021S2-CS2113T-W09-4/tp

# Conflicts:
#	src/main/java/seedu/duke/Constants.java

Emkay16 committed:
Remove redundant import statements

leowxx committed:
Merge pull request #71 from Emkay16/Mabel-userGuide

Update user guide to add expected outcomes to usage examples

leowxx committed:
Merge branch 'master' of https://github.com/AY2021S2-CS2113T-W09-4/tp

leowxx committed:
Merge remote-tracking branch 'origin/master'

Emkay16 committed:
Change details delimiter

fsgmhoward committed:
Merge pull request #70 from Emkay16/Mabel-storage

Add methods to load data from save files

fsgmhoward committed:
Merge remote-tracking branch 'origin/master' into enhancement-1

fsgmhoward committed:
Fix unresolved conflict in Constants.java

brandonfoong committed:
Extract out implementation details into a seperate section

brandonfoong committed:
Re-organize the DG

brandonfoong committed:
Add more details about Exception

fsgmhoward committed:
Handle IO and FNF Exception with custom StorageException

brandonfoong committed:
Make minor changes to formatting

brandonfoong committed:
Update _config.yml

Emkay16 committed:
Merge pull request #73 from fsgmhoward/enhancement-1

Storage and Exception handling

brandonfoong committed:
Fix formatting bugs

brandonfoong committed:
Fix even more formatting issues

fsgmhoward committed:
Merge pull request #75 from fsgmhoward/enhancement-1

Handle IO and FNF Exception with custom StorageException

brandonfoong committed:
Merge branch 'master' into Brandon-Documentation

brandonfoong committed:
Update filename of output file

fsgmhoward committed:
Save data to file after add, delete and record commands

hiongkaihan committed:
Update Java Docs

hiongkaihan committed:
Merge branch 'master' into kaihan-UG

brandonfoong committed:
Add manifest to .gitignore

leowxx committed:
Merge branch 'master' of https://github.com/AY2021S2-CS2113T-W09-4/tp

# Conflicts:
#	src/main/java/seedu/duke/Storage.java
#	src/test/java/seedu/duke/StorageTest.java

fsgmhoward committed:
Accept first argument as file path

- Also fix OOB error at Storage.load() when a patient has no associated
  visit records

Emkay16 committed:
Merge pull request #76 from hiongkaihan/kaihan-UG

Update UGDG

brandonfoong committed:
Update sequence diagram

brandonfoong committed:
Resize diagram

fsgmhoward committed:
Fix JUnit test and IO redirection test

brandonfoong committed:
Add color coding to the diagrams

hiongkaihan committed:
Update JavaDocs and Minor changes to load function

brandonfoong committed:
Make font bigger for visibility

leowxx committed:
Added Storage Component for DG

hiongkaihan committed:
Minor changes

leowxx committed:
Fixed style issues

fsgmhoward committed:
Fix OOB error in prescription, symptom and diagnosis splitting

brandonfoong committed:
Merge branch 'master' and resolve merge conflicts

brandonfoong committed:
Fix formatting

fsgmhoward committed:
Merge pull request #74 from brandonfoong/Brandon-Documentation

Improve documentation

fsgmhoward committed:
Merge pull request #78 from hiongkaihan/kaihan-JavaDocs

Update JavaDocs and Minor changes to load function

leowxx committed:
Merge branch 'master' into XiaoXuan-DG

Emkay16 committed:
Merge pull request #79 from leowxx/XiaoXuan-DG

Added Storage Component to DG

hiongkaihan committed:
Merge branch 'master' into storage-1

hiongkaihan committed:
Update .gitignore

Emkay16 committed:
Merge pull request #77 from fsgmhoward/storage-1

Add triggers to save to file when data changes & bug fixes

Emkay16 committed:
Update help command for new delete command

hiongkaihan committed:
Minor fix to ID checker and coding format

brandonfoong committed:
Merge pull request #80 from hiongkaihan/kaihan-IDChecker

Minor fix to ID checker and coding format

Emkay16 committed:
Fix formatting issues in user guide

Emkay16 committed:
Fix typo

Emkay16 committed:
Merge branch 'master' into Mabel-help

fsgmhoward committed:
Merge pull request #82 from Emkay16/Mabel-userGuide

User guide update

Emkay16 committed:
Update text-ui-test EXPECTED.TXT

Emkay16 committed:
Update text-ui-test EXPECTED.TXT

brandonfoong committed:
Merge pull request #81 from Emkay16/mabel-help

Update help command for new delete command

brandonfoong committed:
Add bugfix for empty record command

brandonfoong committed:
Improve code quality

brandonfoong committed:
Add bugfixes for invalid inputs (Ctrl-D or Ctrl-Z)

brandonfoong committed:
Bugfix for empty argument given to add command

brandonfoong committed:
Cleanup whitespace

brandonfoong committed:
Fix checkstyle violations

cloudy3 commented on docs/DeveloperGuide.md:
```diff
@@ -1,34 +1,442 @@
+
+The Sequence Diagram below shows how the components interact with each other for the scenario where the user issues the command `add S1234567D`.
+
+<img src="./images/sequence_diagram_overview.png" width="750">
```
![image](https://user-images.githubusercontent.com/59783150/113377109-f4bf5980-93a5-11eb-8c48-3d14c9b256c9.png)

May want to use something more generic like: input("add < matric number >"), just a minor suggestion.

Also, there is an empty method call from :Ui to :Logic

lihaoyangML replied:
The sequence diagram is clear and straight forward. Not much suggestions here, but I just want to remind you that the return from method calls (those dotted arrows) are optional if they are unimportant. Hence, if you want to make the sequence diagram even simpler, you may consider removing those unimportant return arrows.

cloudy3 commented on docs/DeveloperGuide.md:
```diff
@@ -1,34 +1,442 @@
+Then, we can call the `save(SortedMap<String, Patient> patientData)` method to save the SortedMap into a file whenever changes are made.
+The file is specified by the variable FILE_PATH in Constants class.
+
+The reverse process would be the `load()` method. The file from the specified FILE_PATH is loaded into a String, which is 
+converted to a SortedMap<String, Patient> afterwards.
```
May want to include a sequence diagram explaining how the `load()` method works

lihaoyangML replied:
The explanation is ok, but it may be more clear to use sequence diagram(s) to explain the working of the storage component.

cloudy3 commented on docs/DeveloperGuide.md:
```diff
@@ -1,34 +1,442 @@
+
+This is broken into a few steps:
+
+1. Initialize an empty hashmap, called `arguments`.
+1. Tokenize using **any number of consecutive white spaces**.
+1. Taken out the first token as command, i.e. `record`. Push it into the hash map using **key** `command`.
+   Create a new empty **list** with default **key** `payload`.
+1. Check if next token starts with `/`. No, so we add it to the list: `list = ['payload']`.
+1. Check if next token starts with `/`. Yes, so we concatenate all tokens in the list to one string use delimiter ` `
+   (empty whitespace). Put it into the hash map using the key `payload`.
+   **Reset the list**, and set new key to `s` (the part after this `/`).
+1. Repeat same process, we have `list = ['coughing,']`
+1. Repeat same process, we have `list = ['coughing,', 'fever']`
+1. Same process, `coughing, fever` is pushed into arguments hash map with key `s`. Reset the list, and new key set to `p`.
+1. ... 
```
This is a long list of lengthy steps, may want to include a sequence diagram to better visualise this?

cloudy3 commented on docs/DeveloperGuide.md:
````diff
@@ -1,34 +1,442 @@
+First is the initialization of this parser. A `Ui` instance and a `Data` instance is passed and stored. This is important as
+these two will be passed to logic components (command classes) later.
+
+Then, we can parse a user-input string by passing it to `parse()`. We use an example of this:
+```
+record 01/05/2021 /s coughing, fever /p panadol Paracetamol 500mg*20
+```
````
May want to include a format for the input string first, before providing an example.
For instance: ```record (dd/MM/YY) /s (symptom), (sickness) /p (medicine dosage)```
A further explanation on the key words like ```record```, ```/s``` and ```/p``` would be nice too

cloudy3 commented on docs/DeveloperGuide.md:
````diff
@@ -1,34 +1,442 @@
+> record 01/05/2021 /s coughing,                 fever
+> ```
+
+### Exception Handling
````
May want to include a class diagram to better visualise the inheritance

cloudy3 reviewed (commented):
Overall very well done! Very in-depth and easy to understand explanations. However, there is a lack of sequence diagrams to better visualise the complex methods.

lihaoyangML commented on docs/DeveloperGuide.md:
```diff
@@ -1,34 +1,442 @@
+
+### Logic Component
+
+<img src="./images/class_diagram_logic.png" width="750">
```
This illustrates clearly on how the logic component works. If time permits, you may want to include more sequence diagrams to explain how different commands work.

lihaoyangML commented on docs/DeveloperGuide.md:
```diff
@@ -1,34 +1,442 @@
+
+### Model Component
+
+<img src="./images/class_diagram_model.png" width="200">
```
The class diagram is simple and clear. Just for your consideration: Would it be better to have more detail for each class in the class diagram (include the variables and methods, along with whether they are private or public...)?

lihaoyangML reviewed (commented):
Overall good work! The explanation and diagrams and simple and clear. I just have some minor suggestions for your consideration.

hiongkaihan committed:
Add exceptions for delete command and minor changes to UG

hiongkaihan committed:
Minor fix

hiongkaihan committed:
Merge pull request #83 from brandonfoong/Brandon-bugfixes

Add bugfixes

hiongkaihan committed:
Merge branch 'master' into kaihan-Exceptions

hiongkaihan committed:
Update InvalidInputException.java

hiongkaihan committed:
Merge pull request #84 from hiongkaihan/kaihan-Exceptions

Add exceptions for delete command and minor changes to UG

fsgmhoward committed:
Add a few notes to UG (#108, #118)

fsgmhoward committed:
Enforce strict parsing of dates entered for `record` command (#99)

fsgmhoward committed:
Forbid record for future dates (#100, #115)

fsgmhoward committed:
Preserve previous loaded patient if load command fails (#106)

fsgmhoward committed:
Improve invalid/empty NRIC error message for `load` command (#107)

fsgmhoward committed:
Use LinkedHashSet to replace ArrayList for details in Record (#116)

- This prevents duplicate details in the same record while order of
  entering is preserved

fsgmhoward committed:
Enable case-insensitive command parsing

- Also add test cases for Parser

fsgmhoward committed:
Improved test coverage for Parser and LoadCommand

fsgmhoward committed:
Update all wrong NRIC numbers, fix IO redirection test output

brandonfoong committed:
Merge pull request #122 from fsgmhoward/master

Multiple fixes to bugs mentioned in PE-D

brandonfoong committed:
Extract the parseDate method into Common

brandonfoong committed:
Add more test cases for record

brandonfoong committed:
Add more test coverage to retrieve and improve code quality

brandonfoong committed:
Add additional checks to prevent invalid inputs
For example, `record /s ` should recognize the symptom as empty and not execute the command

brandonfoong committed:
Fix program logic and add more test cases

Emkay16 committed:
Change help command message to a command list

leowxx committed:
Fixed error message

leowxx committed:
Updated expected.txt

brandonfoong committed:
Rename unit test

brandonfoong committed:
Extract parseDate method into Common and create a formatDate function

brandonfoong committed:
Remove hardcoded whitespace in Constants
printError will handle tokenization instead
Re-organize Constants

brandonfoong committed:
Change printError to handle the new error format

brandonfoong committed:
Update all tests to use assertDoesNotThrow instead of a try-catch block

brandonfoong committed:
Update text-ui-tests

leowxx committed:
Modified error message

Emkay16 committed:
Merge pull request #124 from brandonfoong/Brandon-unittests

Improve Unit Tests

leowxx committed:
Fixed error messages

leowxx committed:
Bug fixes

leowxx committed:
Merge branch 'master' into XiaoXuan-BugFix

leowxx committed:
Update BaseException.java

brandonfoong committed:
Test re-numbering of lists

brandonfoong committed:
Test numbered list formatting

brandonfoong committed:
Update info section of the UG

brandonfoong committed:
Update formatting

brandonfoong committed:
Test nested comment blocks

brandonfoong committed:
Minor edits to UG

brandonfoong committed:
Change image formatting

brandonfoong committed:
Update quickstart

brandonfoong committed:
Update features in UG

brandonfoong committed:
Reformat numbered list

brandonfoong committed:
Update Features section

brandonfoong committed:
Fix list styling

brandonfoong committed:
Update command format

brandonfoong committed:
Fix formatting

brandonfoong committed:
Update documentation for help

brandonfoong committed:
Improve documentation for delete

brandonfoong committed:
More changes to the UG

brandonfoong committed:
Standardize argument names and remove whitespace/order of inputs from the TOC

brandonfoong committed:
Minor syntax fixes

brandonfoong committed:
Minor syntax fixes

fsgmhoward committed:
Revert changes for ParserTest

brandonfoong committed:
Use dd/mm/yyyy formatted date in DeleteCommand

brandonfoong committed:
Make changes to RecordCommand and add a DataException class to handle exceptions in the Data class

brandonfoong committed:
Reduce coupling in the RecordCommand class and update tests accordingly

brandonfoong committed:
Fix date format in DeleteCommandTest

fsgmhoward committed:
Merge pull request #125 from leowxx/XiaoXuan-BugFix

Fixed error message

brandonfoong committed:
Improve OOP of RetrieveCommand and update tests

brandonfoong committed:
Fix program logic and update text ui tests

fsgmhoward committed:
Add PPP for fsgmhoward, multiple fixes to docs

- Fix problems with Table of Content in DG
- Fix typos and format problems in both UG and DG
- Update DG on Exception and common classes
- Remove unused import for RetrieveCommand

fsgmhoward committed:
Merge branch 'master' of https://github.com/AY2021S2-CS2113T-W09-4/tp

brandonfoong committed:
Improve OOP-ness of DeleteCommand and update tests

brandonfoong committed:
Improve OOP-ness of LoadCommand and update tests

brandonfoong committed:
Improve OOP-ness of CurrentCommand

brandonfoong committed:
Rename error constants to be consistent with their exception classes

brandonfoong committed:
Merge branch 'master' into Brandon-improveOOP

brandonfoong committed:
Update text ui tests

brandonfoong committed:
Add PPP (format shamelessly stolen :) )

brandonfoong committed:
Fix urls and formatting in PPP

hiongkaihan committed:
Add exceptions and bugfixes

hiongkaihan committed:
Minor fix

hiongkaihan committed:
Minor fix

hiongkaihan committed:
Minor fix

hiongkaihan committed:
Minor fix

fsgmhoward committed:
Fix minor typo for PPP

Emkay16 committed:
Merge branch 'master' into Mabel-help

Emkay16 committed:
Fix issue where duplicated arguments results in duplicated output in help command

Emkay16 committed:
Update EXPEXTED.TXT to reflect changes in help command output

Emkay16 committed:
Merge pull request #126 from brandonfoong/Brandon-Documentation

Improve documentation and add PPP

brandonfoong committed:
Update UserGuide.md

Fix css numbering

leowxx committed:
Merge branch 'master' into master

leowxx committed:
Merge pull request #128 from fsgmhoward/master

PPP and multiple fixes for UGDG

Emkay16 committed:
Update information message about help command

hiongkaihan committed:
Minor fixes

hiongkaihan committed:
Minor fix

brandonfoong committed:
Merge pull request #129 from hiongkaihan/kaihan-Bugfixes

Add exceptions and bugfixes

fsgmhoward committed:
Add filter for save file delimiters, fix new line problem for exceptions

Emkay16 committed:
Update help message

Emkay16 committed:
Update EXPECTED.TXT to reflect changes to help message

brandonfoong committed:
Merge branch 'master' into Brandon-improveOOP

brandonfoong committed:
Merge changes from upstream master

Emkay16 committed:
Add author tag to Storage.java

brandonfoong committed:
Update text ui tests

fsgmhoward committed:
Add TC for forbidden string filter, update UGDG

fsgmhoward committed:
Add CSS for numbering of TOC in DG

brandonfoong committed:
Test changing emojis to follow GitHub flavoured Markdown standards

brandonfoong committed:
Replace all emojis with GFM equivalents

fsgmhoward committed:
Fix ParserTest

brandonfoong committed:
Add jemoji plugin to render emoji properly

brandonfoong committed:
Update UG to be consistent with invalid chars

fsgmhoward committed:
Merge pull request #132 from brandonfoong/Brandon-Documentation

Fix emojis

fsgmhoward committed:
Merge pull request #130 from Emkay16/Mabel-help

Fixes for help command

leowxx committed:
Merge branch 'master' into delimiter-filter

leowxx committed:
Merge pull request #131 from fsgmhoward/delimiter-filter

Delimiter filter and UGDG update

brandonfoong committed:
Merge branch 'master' into Brandon-improveOOP

brandonfoong committed:
Fix code and text ui tests

fsgmhoward committed:
Add Command Summary for DG, and split delete command for UG

fsgmhoward committed:
Update link for Command Summary for DG

Emkay16 committed:
Merge pull request #133 from fsgmhoward/ugdg-1

Add Command Summary for DG, and split delete command for UG

leowxx committed:
Added PPP

Emkay16 committed:
Add PPP

Emkay16 committed:
Merge branch 'master' into Mabel-PPP

Emkay16 committed:
Update AboutUs.md

brandonfoong committed:
Extract checkdigits as constants

brandonfoong committed:
Update sequence diagram and DG

brandonfoong committed:
Merge branch 'master' into Brandon-Documentation

brandonfoong committed:
Update DG

brandonfoong committed:
Merge branch 'master' into Brandon-Documentation

# Conflicts:
#	docs/DeveloperGuide.md

brandonfoong committed:
Merge latest changes from master and add writeup for design of the Data class

brandonfoong committed:
Change site title

brandonfoong committed:
Add hyperlinks for easier navigation

brandonfoong committed:
Change page title

brandonfoong committed:
Remove page title and rename sections

brandonfoong committed:
Remove exception from archi diagram

brandonfoong committed:
Add class diagram for exceptions

brandonfoong committed:
Fix numbering and images

brandonfoong committed:
Add writeup for exceptions

brandonfoong committed:
Add writeup to exception

Emkay16 committed:
Fix broken links on User Guide

brandonfoong committed:
Add seq diagram for "add" command

brandonfoong committed:
Change sequence diagram sizes

brandonfoong committed:
Add seq diagram for load

brandonfoong committed:
Update TOC

brandonfoong committed:
Fix code indentation

brandonfoong committed:
Correct diagrams to data instead of model

hiongkaihan committed:
Added PPP

brandonfoong committed:
Add details for record and retrieve

brandonfoong committed:
Remove extra loading patients, update TOC

brandonfoong committed:
Replace emojis with GFM equivalents

leowxx committed:
Merge pull request #136 from hiongkaihan/kaihan-PPP

Added PPP

leowxx committed:
Merge pull request #135 from Emkay16/Mabel-PPP

Add PPP and update AboutUs.md

hiongkaihan committed:
Merge pull request #134 from leowxx/XiaoXuan-PPP

Added PPP

Emkay16 committed:
Updated links

brandonfoong committed:
Update all instances of tp.jar to PatientManager.jar

brandonfoong committed:
Fix list numbering in UG

brandonfoong committed:
Fix broken links in PPP, minor grammar changes

fsgmhoward committed:
Merge pull request #138 from Emkay16/Mabel-DeveloperGuide

Update command summary for UG and DG

brandonfoong committed:
Merge branch 'master' into Brandon-Documentation

fsgmhoward committed:
Merge pull request #137 from brandonfoong/Brandon-Documentation

Update DG

fsgmhoward committed:
Merge branch 'master' into Brandon-improveOOP

fsgmhoward committed:
Merge pull request #127 from brandonfoong/Brandon-improveOOP

Improve OOP-ness of the code

Emkay16 committed:
Update help messages

Emkay16 committed:
Update EXPECTED.TXT accordingly

brandonfoong committed:
Update DeveloperGuide.md

Quick fixes to expected outputs

brandonfoong committed:
Merge pull request #139 from Emkay16/Mabel-help

Update help messages

fsgmhoward committed:
Update PPP to fix link

fsgmhoward committed:
Merge pull request #140 from fsgmhoward/master

Update PPP to fix link

brandonfoong committed:
Fix add bug and add documentation for getCurrentPatientDetails

brandonfoong committed:
Merge branch 'master' of github.com:brandonfoong/tp

brandonfoong committed:
Fix coding style issues

Emkay16 committed:
Update links

Emkay16 committed:
Update wrong links

Emkay16 committed:
Merge pull request #142 from Emkay16/master

Update links

Emkay16 committed:
Update links

Emkay16 committed:
Remove unhidden comment
